### PR TITLE
OpenZFS 7090 - zfs should throttle allocations

### PR DESCRIPTION
**Pull request by don-brady**
Authored by: George Wilson <george.wilson@delphix.com>
Reviewed by: Alex Reece <alex@delphix.com>
Reviewed by: Christopher Siden <christopher.siden@delphix.com>
Reviewed by: Dan Kimmel <dan.kimmel@delphix.com>
Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: Paul Dagnelie <paul.dagnelie@delphix.com>
Reviewed by: Prakash Surya <prakash.surya@delphix.com>
Reviewed by: Sebastien Roy <sebastien.roy@delphix.com>

Approved by: Matthew Ahrens <mahrens@delphix.com>

When write I/Os are issued, they are issued in block order but the ZIO
pipeline will drive them asynchronously through the allocation stage
which can result in blocks being allocated out-of-order. It would be
nice to preserve as much of the logical order as possible.

In addition, the allocations are equally scattered across all top-level
VDEVs but not all top-level VDEVs are created equally. The pipeline
should be able to detect devices that are more capable of handling
allocations and should allocate more blocks to those devices. This
allows for dynamic allocation distribution when devices are imbalanced
as fuller devices will tend to be slower than empty devices.

The change includes a new pool-wide allocation queue which would
throttle and order allocations in the ZIO pipeline. The queue would be
ordered by issued time and offset and would provide an initial amount of
allocation of work to each top-level vdev. The allocation logic utilizes
a reservation system to reserve allocations that will be performed by
the allocator. Once an allocation is successfully completed it's
scheduled on a given top-level vdev. Each top-level vdev maintains a
maximum number of allocations that it can handle (mg_alloc_queue_depth).
The pool-wide reserved allocations (top-levels * mg_alloc_queue_depth)
are distributed across the top-level vdevs metaslab groups and round
robin across all eligible metaslab groups to distribute the work. As
top-levels complete their work, they receive additional work from the
pool-wide allocation queue until the allocation queue is emptied.

Ported-by: Don Brady <don.brady@intel.com>
OpenZFS-issue: https://www.illumos.org/issues/7090
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/4756c3d7

Porting Notes:
	- Maintained minimal stack in zio_done
	- Preserve linux-specific io sizes in zio_write_compress

### CODE REVIEW

File: include/sys/zio_impl.h [hunk 1]
```diff
@@ -24,7 +24,7 @@
  */
 
 /*
- * Copyright (c) 2013 by Delphix. All rights reserved.
+ * Copyright (c) 2012, 2015 by Delphix. All rights reserved.
```
Comment (rlaager) [inline include/sys/zio_impl.h]: This is dropping the 2013 copyright. Was that intentional, and if so, why?

File: module/zfs/vdev_mirror.c [hunk 2]
```diff
@@ -24,7 +24,7 @@
  */
 
 /*
- * Copyright (c) 2012, 2014 by Delphix. All rights reserved.
+ * Copyright (c) 2012, 2015 by Delphix. All rights reserved.
```
Comment (rlaager) [inline module/zfs/vdev_mirror.c]: Likewise, this drops a 2014 copyright.

File: include/sys/zio_impl.h (hunk 1 above)
Comment (kernelOfTruth) [inline include/sys/zio_impl.h]: Not certain how it's in this case but I also handled it that way - if e.g. upstream commits from Illumos did it

@behlendorf @ahrens  would be good to know how the general policy is on how to proceed on these cases

Comment (rlaager) [inline include/sys/zio_impl.h]: Can you clarify who "We" is (Delphix, Illumos, OpenZFS) in that policy? And you don't want to connect those years with a dash (e.g. 2012-2015)?

Comment (ahrens) [inline include/sys/zio_impl.h]: "We" is Delphix, the copyright holder.

Comment (don-brady) [inline include/sys/zio_impl.h]: good observation.  This change was directly from Delpix in their commit to Open ZFS.  https://github.com/illumos/illumos-gate/commit/4756c3d

Comment (rlaager) [inline include/sys/zio_impl.h]: Thanks. I see no reason to object to a copyright holder changing their own copyright line.

File: module/zfs/spa.c [hunk 3]
```diff
@@ -6469,6 +6478,10 @@ spa_sync(spa_t *spa, uint64_t txg)
 	spa->spa_syncing_txg = txg;
 	spa->spa_sync_pass = 0;
 
+	mutex_enter(&spa->spa_alloc_lock);
+	VERIFY0(avl_numnodes(&spa->spa_alloc_tree));
+	mutex_exit(&spa->spa_alloc_lock);
```
Comment (behlendorf) [inline module/zfs/spa.c]: This feels pretty heavy handed for a release build.  I'd be tempted to wrap it in ZFS_DEBUG but there's no point diverging here unless it's actually a problem.

File: module/zfs/spa.c [hunk 4]
```diff
@@ -6676,6 +6721,10 @@ spa_sync(spa_t *spa, uint64_t txg)
 
 	dsl_pool_sync_done(dp, txg);
 
+	mutex_enter(&spa->spa_alloc_lock);
+	VERIFY0(avl_numnodes(&spa->spa_alloc_tree));
+	mutex_exit(&spa->spa_alloc_lock);
```
Comment (behlendorf) [inline module/zfs/spa.c]: Same.

File: module/zfs/zio.c [hunk 5]
```diff
@@ -522,6 +528,30 @@ zio_inherit_child_errors(zio_t *zio, enum zio_child c)
...
+		return (1);
+
+	if (z1 < z2)
+		return (-1);
+	if (z1 > z2)
+		return (1);
+
+	return (0);
+}
+
```
Comment (behlendorf) [inline module/zfs/zio.c]: For Linux we optimized the other AVL tree comparator functions in ee36c709c3d5f7040e1bd11f5c75318aa03e789f.  We should give this one the same treatment.

File: module/zfs/vdev_queue.c [hunk 6]
```diff
@@ -171,6 +172,23 @@ int zfs_vdev_aggregation_limit = SPA_OLD_MAXBLOCKSIZE;
...
+ * then allocator will stop allocating blocks on that top-level device.
+ * The default kernel setting is 1000% which will yield 100 allocations per
+ * device. For userland testing, the default setting is 300% which equates
+ * to 30 allocations per device.
+ */
+#ifdef _KERNEL
+int zfs_vdev_queue_depth_pct = 1000;
+#else
+int zfs_vdev_queue_depth_pct = 300;
+#endif
```
Comment (don-brady) [inline module/zfs/vdev_queue.c]: Got it. Will add them.

File: module/zfs/spa.c [hunk 7]
```diff
@@ -113,7 +113,11 @@ typedef struct zio_taskq_info {
 } zio_taskq_info_t;
 
 static const char *const zio_taskq_types[ZIO_TASKQ_TYPES] = {
+#ifdef __linux__
 	"iss", "iss_h", "int", "int_h"
+#else
+	"issue", "issue_high", "intr", "intr_high"
+#endif
```
Comment (don-brady) [inline module/zfs/spa.c]: The change in zio.c for `zio_type_name` was the motivation. I had to go chase down why they didn't match openzfs.  I'd be happy to just make it a comment for future integrators.

**Pull request merged**